File: src/game/monsterCombat.ts
export type MonsterCombatState = {
  maxHp: number
  currentHp: number
  contactDamage: number
}

type CreateMonsterCombatStateOptions = {
  hpMultiplier?: number
  damageMultiplier?: number
}

const MONSTER_BASE_HP = 10
const MONSTER_HP_PER_LEVEL = 2

export const createMonsterCombatState = (
  monsterLevel = 1,
  options: CreateMonsterCombatStateOptions = {}
): MonsterCombatState => {
  const level = Math.max(1, Math.floor(monsterLevel))
  const hpMultiplier = Math.max(1, options.hpMultiplier ?? 1)
  const damageMultiplier = Math.max(1, options.damageMultiplier ?? 1)
  const maxHp = (MONSTER_BASE_HP + level * MONSTER_HP_PER_LEVEL) * hpMultiplier

  return {
    maxHp,
    currentHp: maxHp,
    contactDamage:
      Math.max(1, Math.ceil(level / 2)) * damageMultiplier
  }
}

export const applyMonsterDamage = (
  combatState: MonsterCombatState,
  damage: number
): MonsterCombatState => {
  const nextDamage = Math.max(0, Math.floor(damage))

  if (nextDamage === 0 || combatState.currentHp === 0) {
    return combatState
  }

  const nextCurrentHp = Math.max(0, combatState.currentHp - nextDamage)

  if (nextCurrentHp === combatState.currentHp) {
    return combatState
  }

  return {
    ...combatState,
    currentHp: nextCurrentHp
  }
}

export const isMonsterDefeated = (
  combatState: MonsterCombatState
): boolean => combatState.currentHp <= 0

File: src/games/my-sample-rpg/assets/lua/monster-combat.lua
-- 몬스터 전투 상태 규칙 (TS monsterCombat.ts에서 Lua로 변환).
-- 순수 산술 — 호스트가 전역 함수를 원시값 인자로 호출해 number를 받는다.
-- monsterCombat.ts와 동일: level = max(1, floor(monster_level)).

local MONSTER_BASE_HP = 10
local MONSTER_HP_PER_LEVEL = 2

function monster_combat_create(monster_level, hp_multiplier, damage_multiplier)
  local level = math.max(1, math.floor(monster_level))
  local hp_mult = math.max(1, hp_multiplier)
  local dmg_mult = math.max(1, damage_multiplier)
  local max_hp = (MONSTER_BASE_HP + level * MONSTER_HP_PER_LEVEL) * hp_mult
  local contact_damage = math.max(1, math.ceil(level / 2)) * dmg_mult

  return max_hp, max_hp, contact_damage
end

function monster_combat_apply_damage(current_hp, damage)
  local next_damage = math.max(0, math.floor(damage))

  if next_damage == 0 or current_hp == 0 then
    return current_hp
  end

  return math.max(0, current_hp - next_damage)
end

function monster_combat_is_defeated(current_hp)
  if current_hp <= 0 then
    return 1
  end
  return 0
end
